components/core/src/clp/StreamingReader.hpp

```diff
@@ -0,0 +1,411 @@
+ * from a pre-signed S3 URL). When open an URL< a daemon thread will be created to fetch data in the
+ * background using libcurl. This class guarantees the synchronization between the read operations
+ * and data streaming (fetching). However, it is assumed that both the reader and the data fetcher
+ * are single threaded.
+ */
+class StreamingReader : public ReaderInterface {
+public:
+    using BufferView = std::span<char>;
+
+    // Types
+    /**
+     * This class defines the exception thrown by the streaming reader.
+     */
+    class OperationFailed : public TraceableException {
+    public:
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return "StreamingReader operation failed.";
+        }
+    };
+
+    /**
+     * This enum defines the states of the reader.
+     * NotInit: The reader is not initialized with a source URL.
+     * Failed: The reader has failed to stream data from the source URL.
+     * InProgress: The reader is fetching data from the source URL.
+     * Finished: The streaming of the given source URL has been accomplished.
+     */
+    enum class StatusCode : uint8_t {
+        NotInit = 0,
+        Failed,
+        InProgress,
+        Finished
+    };
+
+    /**
+     * These static members defines the default buffer size and the number of pools.
+     */
+    static constexpr size_t cDefaultBufferPoolSize{8};
+    static constexpr size_t cDefaultBufferSize{4096};
+
+    /**
+     * These static members defines the minimum allowed buffer size and the number of pools.
+     */
+    static constexpr size_t cMinimalBufferSize{512};
+    static constexpr size_t cMinimalBufferPoolSize{2};
+
+    /**
+     * These static members defines the default connection timeout and operation timeout.
+     * Please refer to the libcurl documentation for more details:
+     * ConnectionTimeout: https://curl.se/libcurl/c/CURLOPT_CONNECTTIMEOUT.html
+     * OperationTimeout: https://curl.se/libcurl/c/CURLOPT_TIMEOUT.html
+     */
+    static constexpr uint32_t cDefaultConnectionTimeout{10};
+    static constexpr uint32_t cDefaultOperationTimeout{45};
+
+    /**
+     * Initializes the underlying libcurl functionalities globally. It should be called before
+     * opening a source URL for data fetching.
+     * @return ErrorCode_Success on success.
+     * @return ErrorCode_Failure if the libcurl initialization failed.
+     */
+    [[nodiscard]] static auto global_init() -> ErrorCode;
+
+    /**
+     * Releases the globally initialized libcurl resources.
+     */
+    static auto global_cleanup() -> void;
+
+    /**
+     * Constructor.
+     * @param buffer_pool_size Total number of buffers available for fetching.
+     * @param buffer_size The size of each data buffer.
+     */
+    StreamingReader(
```

Comment:
```suggestion
    explicit StreamingReader(
```
according to clang-tidy